soc/intel/intel_adsp/tools/cavstool.py

```diff
@@ -461,6 +495,130 @@ def load_firmware(fw_file):
     sd.CTL |= 1
     log.info(f"cAVS firmware load complete")
 
+def load_firmware_ace(fw_file):
+    try:
+        fw_bytes = open(fw_file, "rb").read()
+        # Resize fw_bytes for MTL
+        if len(fw_bytes) < 512 * 1024:
+            fw_bytes += b'\x00' * (512 * 1024 - len(fw_bytes))
+    except Exception as e:
+        log.error(f"Could not read firmware file: `{fw_file}'")
+        log.error(e)
+        sys.exit(1)
+
+    (magic, sz) = struct.unpack("4sI", fw_bytes[0:8])
+    if magic == b'$AE1':
+        log.info(f"Trimming {sz} bytes of extended manifest")
+        fw_bytes = fw_bytes[sz:len(fw_bytes)]
+
+    # This actually means "enable access to BAR4 registers"!
+    hda.PPCTL |= (1 << 30) # GPROCEN, "global processing enable"
+
+    log.info("Resetting HDA device")
+    hda.GCTL = 0
+    while hda.GCTL & 1: pass
+    hda.GCTL = 1
+    while not hda.GCTL & 1: pass
+
+    log.info("Turning of DSP subsystem")
+    dsp.HFDSSCS &= ~(1 << 16) # clear SPA bit
+    time.sleep(0.002)
+    # wait for CPA bit clear
+    while dsp.HFDSSCS & (1 << 24):
+        log.info("Waiting for DSP subsystem power off")
+        time.sleep(0.1)
+
+    log.info("Turning on DSP subsystem")
+    dsp.HFDSSCS |= (1 << 16) # set SPA bit
+    time.sleep(0.002) # needed as the CPA bit may be unstable
+    # wait for CPA bit
+    while not dsp.HFDSSCS & (1 << 24):
+        log.info("Waiting for DSP subsystem power on")
+        time.sleep(0.1)
+
+    log.info("Turning on Domain0")
+    dsp.HFPWRCTL |= 0x1 # set SPA bit
+    time.sleep(0.002) # needed as the CPA bit may be unstable
+    # wait for CPA bit
+    while not dsp.HFPWRSTS & 0x1:
+        log.info("Waiting for DSP domain0 power on")
+        time.sleep(0.1)
+
+    log.info("Turning off Primary Core")
+    dsp.DSP2CXCTL_PRIMARY &= ~(0x1) # clear SPA
+    time.sleep(0.002) # wait for CPA settlement
+    while dsp.DSP2CXCTL_PRIMARY & (1 << 8):
+        log.info("Waiting for DSP primary core power off")
+        time.sleep(0.1)
+
+
+    log.info(f"Configuring HDA stream {hda_ostream_id} to transfer firmware image")
+    (buf_list_addr, num_bufs) = setup_dma_mem(fw_bytes)
+    sd.CTL = 1
+    while (sd.CTL & 1) == 0: pass
+    sd.CTL = 0
+    while (sd.CTL & 1) == 1: pass
+    sd.CTL |= (1 << 20) # Set stream ID to anything non-zero
+    sd.BDPU = (buf_list_addr >> 32) & 0xffffffff
+    sd.BDPL = buf_list_addr & 0xffffffff
+    sd.CBL = len(fw_bytes)
+    sd.LVI = num_bufs - 1
+    hda.PPCTL |= (1 << hda_ostream_id)
+
+    # SPIB ("Software Position In Buffer") is an Intel HDA extension
+    # that puts a transfer boundary into the stream beyond which the
+    # other side will not read.  The ROM wants to poll on a "buffer
+    # full" bit on the other side that only works with this enabled.
+    hda.SPBFCTL |= (1 << hda_ostream_id)
+    hda.SD_SPIB = len(fw_bytes)
+
+
```

Comment:
Another double line

Reply:
D'oh, one was existing double-line I copied verbatim, but it seems the style caught on and added a few of my own. Double lines now removed.